### PR TITLE
Report when ibv_fork_init() is not needed

### PR DESCRIPTION
**Pull request by galpress**
Followup from #883, this series identifies kernel with copy-on-fork support and reports that through the recently introduced ibv_is_fork_initialized() verb.

Kernel submission:
v1: https://lore.kernel.org/linux-rdma/20210405114722.98904-1-galpress@amazon.com/
v2: https://lore.kernel.org/linux-rdma/20210407101606.80737-1-galpress@amazon.com/
v3: https://lore.kernel.org/linux-rdma/20210412064150.40064-1-galpress@amazon.com/
v4: https://lore.kernel.org/linux-rdma/20210418121025.66849-1-galpress@amazon.com/

### CODE REVIEW

File: libibverbs/ibdev_nl.c [hunk 1]
```diff
@@ -217,3 +217,41 @@ int find_sysfs_devs_nl(struct list_head *tmp_sysfs_dev_list)
...
+	if (!tb[RDMA_NLDEV_SYS_ATTR_COPY_ON_FORK]) {
+		*(uint8_t *)data = 0;
+		return NL_OK;
+	}
+
+	*(uint8_t *)data = nla_get_u8(tb[RDMA_NLDEV_SYS_ATTR_COPY_ON_FORK]);
+	return NL_OK;
+}
+
+bool get_copy_on_fork(void)
```
Comment (mark-bloch) [inline libibverbs/ibdev_nl.c:241]: Do you except it to change during the RDMA device's lifetime (kernel patch always reports it as supported)?
Seems kind of wasteful to create a socket -> talk with the kernel -> close socket
on every call to ibv_is_fork_initialized() if the result never changes.

Maybe store it (like we do with ibdev_idx and such) while in find_sysfs_devs_nl_cb()?

File: libibverbs/ibdev_nl.c [hunk 2]
```diff
@@ -217,3 +217,41 @@ int find_sysfs_devs_nl(struct list_head *tmp_sysfs_dev_list)
...
+{
+	struct nl_sock *nl;
+	uint8_t cof;
+
+	nl = rdmanl_socket_alloc();
+	if (!nl)
+		return false;
+
+	if (rdmanl_get_copy_on_fork(nl, get_copy_on_fork_cb, &cof)) {
+		nl_socket_free(nl);
```
[the remark this answers is not in the document]
Comment (galpress) [reply libibverbs/ibdev_nl.c:251]: Hey Mark 😃 
Sure.

File: libibverbs/ibdev_nl.c (hunk 1 above)
Comment (galpress) [reply libibverbs/ibdev_nl.c:241]: No, it's not expected to change.
On the other hand, I don't expect this function to get called more than once/twice on init time.

**Comment by galpress** [comment]
I'm trying to think whether we should change the behavior of ibv_fork_init() in case cof is supported (do nothing and return success? return an error?).
Curious to hear other people thoughts about it.

**Comment by galpress** [comment]
Gentle reminder.

**Review by rleon — APPROVED** [review]

**Comment by galpress** [comment]
Thanks Leon!

> I'm trying to think whether we should change the behavior of ibv_fork_init() in case cof is supported (do nothing and return success?).
> Curious to hear other people thoughts about it.

This questions remains open.

**Comment by rleon** [comment]
> Thanks Leon!
> 
> > I'm trying to think whether we should change the behavior of ibv_fork_init() in case cof is supported (do nothing and return success?).
> > Curious to hear other people thoughts about it.
> 
> This question remains open.

I don't know what such change will give us.

Thanks

**Comment by galpress** [comment]
This could allow existing apps that don't use ibv_is_fork_initialized() to benefit from the kernel's cof, even if they called ibv_fork_init().

**Comment by rleon** [comment]
This answer is too vague.

Do you expect any performance improvement? Will it improve any runtime/initialization parameter significantly?

**Comment by galpress** [comment]
It should improve performance for cases where registration is done in data path such as MPI, as it will skip the tree traversal and kernel calls.

I had some numbers about the fork overhead with huge pages back at: https://github.com/linux-rdma/rdma-core/pull/767.

**Comment by krehm** [comment]
In my case it's not a performance question, it's a way to prevent segfaults when users specify environment variables that enable ibv_fork_init().

**Pull request merged**